File: code/rustdb-core/src/transaction/wal.rs
use std::fs::{File, OpenOptions};
use std::io::{Write, Read};
use std::path::Path;

pub const WAL_PATH: &str = "rustdb.wal";

/// WAL 자동 체크포인트 임계값 (512KB)
pub const AUTO_CHECKPOINT_BYTES: u64 = 512 * 1024;

/// WAL 레코드 op 코드
#[repr(u8)]
#[derive(Debug, Clone, PartialEq)]
pub enum WalOp {
    Insert    = 0x01,
    Update    = 0x02,
    Delete    = 0x03,
    Commit    = 0xFF,
    Rollback  = 0xFE,
    Checkpoint= 0xFD,
}

impl WalOp {
    pub fn from_u8(v: u8) -> Option<WalOp> {
        match v {
            0x01 => Some(WalOp::Insert),
            0x02 => Some(WalOp::Update),
            0x03 => Some(WalOp::Delete),
            0xFF => Some(WalOp::Commit),
            0xFE => Some(WalOp::Rollback),
            0xFD => Some(WalOp::Checkpoint),
            _    => None,
        }
    }
}

/// 바이너리 WAL 레코드
#[derive(Debug, Clone)]
pub struct WalRecord {
    pub op:         WalOp,
    pub table_name: String,
    pub key:        String,
    pub data:       String, // JSON
}

pub struct WalManager {
    path: String,
}

impl WalManager {
    pub fn new() -> Self {
        WalManager { path: WAL_PATH.to_string() }
    }

    pub fn new_with_dir(dir: &str) -> Self {
        WalManager { path: format!("{}/rustdb.wal", dir) }
    }

    /// 레코드 바이너리 인코딩
    /// [ op(1) | table_len(4) | table(n) | key_len(4) | key(n) | data_len(4) | data(n) ]
    fn encode(record: &WalRecord) -> Vec<u8> {
        let mut buf = Vec::new();
        let table_bytes = record.table_name.as_bytes();
        let key_bytes   = record.key.as_bytes();
        let data_bytes  = record.data.as_bytes();

        buf.push(record.op.clone() as u8);
        buf.extend_from_slice(&(table_bytes.len() as u32).to_le_bytes());
        buf.extend_from_slice(table_bytes);
        buf.extend_from_slice(&(key_bytes.len() as u32).to_le_bytes());
        buf.extend_from_slice(key_bytes);
        buf.extend_from_slice(&(data_bytes.len() as u32).to_le_bytes());
        buf.extend_from_slice(data_bytes);
        buf
    }

    /// 레코드 바이너리 디코딩
    fn decode(buf: &[u8], pos: &mut usize) -> Option<WalRecord> {
        if *pos >= buf.len() { return None; }

        let op = WalOp::from_u8(buf[*pos])?;
        *pos += 1;

        let table_name = Self::read_string(buf, pos)?;
        let key        = Self::read_string(buf, pos)?;
        let data       = Self::read_string(buf, pos)?;

        Some(WalRecord { op, table_name, key, data })
    }

    fn read_string(buf: &[u8], pos: &mut usize) -> Option<String> {
        if *pos + 4 > buf.len() { return None; }
        let len = u32::from_le_bytes(buf[*pos..*pos+4].try_into().ok()?) as usize;
        *pos += 4;
        if *pos + len > buf.len() { return None; }
        let s = String::from_utf8(buf[*pos..*pos+len].to_vec()).ok()?;
        *pos += len;
        Some(s)
    }

    /// 인코딩된 바이트를 WAL 파일에 기록. sync=true 이면 커널 버퍼 → 디스크 fsync.
    fn write_encoded(&self, encoded: &[u8], sync: bool) {
        let mut file = OpenOptions::new()
            .create(true)
            .append(true)
            .open(&self.path)
            .expect("WAL 파일 열기 실패");
        file.write_all(encoded).expect("WAL 기록 실패");
        if sync {
            file.sync_all().expect("WAL fsync 실패");
        }
    }

    /// WAL에 레코드 기록 (fsync 없음 — 데이터 변경 레코드용)
    pub fn append(&self, record: WalRecord) {
        self.write_encoded(&Self::encode(&record), false);
    }

    pub fn log_insert(&self, table: &str, key: &str, data: &str) {
        self.append(WalRecord {
            op: WalOp::Insert,
            table_name: table.to_string(),
            key: key.to_string(),
            data: data.to_string(),
        });
    }

    pub fn log_update(&self, table: &str, key: &str, data: &str) {
        self.append(WalRecord {
            op: WalOp::Update,
            table_name: table.to_string(),
            key: key.to_string(),
            data: data.to_string(),
        });
    }

    pub fn log_delete(&self, table: &str, key: &str) {
        self.append(WalRecord {
            op: WalOp::Delete,
            table_name: table.to_string(),
            key: key.to_string(),
            data: String::new(),
        });
    }

    /// COMMIT — fsync로 커밋 레코드를 디스크에 영속화 (innodb_flush_log_at_trx_commit=1 동등)
    pub fn log_commit(&self) {
        let record = WalRecord {
            op: WalOp::Commit,
            table_name: String::new(),
            key: String::new(),
            data: String::new(),
        };
        self.write_encoded(&Self::encode(&record), true);
    }

    /// COMMIT 레코드를 기록하되 fsync하지 않음 (Group Commit용).
    /// 호출자가 이후 GroupCommitCoordinator::sync_commit()으로 fsync를 보장해야 한다.
    pub fn log_commit_no_sync(&self) {
        let record = WalRecord {
            op: WalOp::Commit,
            table_name: String::new(),
            key: String::new(),
            data: String::new(),
        };
        self.write_encoded(&Self::encode(&record), false);
    }

    pub fn log_rollback(&self) {
        self.append(WalRecord {
            op: WalOp::Rollback,
            table_name: String::new(),
            key: String::new(),
            data: String::new(),
        });
    }

    /// CHECKPOINT — 버퍼풀 플러시 완료 표시를 디스크에 영속화
    pub fn log_checkpoint(&self) {
        let record = WalRecord {
            op: WalOp::Checkpoint,
            table_name: String::new(),
            key: String::new(),
            data: String::new(),
        };
        self.write_encoded(&Self::encode(&record), true);
    }

    /// WAL 전체 읽기 (복구용)
    pub fn read_all(&self) -> Vec<WalRecord> {
        if !Path::new(&self.path).exists() {
            return vec![];
        }
        let mut file = File::open(&self.path).expect("WAL 읽기 실패");
        let mut buf = Vec::new();
        file.read_to_end(&mut buf).expect("WAL 읽기 실패");

        let mut records = Vec::new();
        let mut pos = 0;
        while let Some(record) = Self::decode(&buf, &mut pos) {
            records.push(record);
        }
        records
    }

    /// WAL 파일 삭제 (체크포인트 후)
    pub fn clear(&self) {
        if Path::new(&self.path).exists() {
            std::fs::remove_file(&self.path).ok();
        }
    }

    /// WAL 파일 크기 (bytes)
    pub fn file_size(&self) -> u64 {
        std::fs::metadata(&self.path).map(|m| m.len()).unwrap_or(0)
    }

    /// 마지막 CHECKPOINT 이후 레코드만 남기고 WAL을 재작성한다.
    /// CHECKPOINT 레코드 자체는 포함시킨다 (복구 시작점 표시).
    pub fn truncate_to_last_checkpoint(&self) {
        let records = self.read_all();
        if records.is_empty() { return; }

        // 마지막 CHECKPOINT 위치 찾기
        let last_cp = records.iter().rposition(|r| r.op == WalOp::Checkpoint);
        let Some(cp_idx) = last_cp else { return };

        // CHECKPOINT 이전 레코드는 이미 디스크에 반영됐으므로 제거
        let remaining = &records[cp_idx..];

        // CHECKPOINT 레코드 하나만 남은 경우 WAL 전체 초기화
        if remaining.len() <= 1 {
            self.clear();
            return;
        }

        // 나머지 레코드로 WAL 재작성
        let mut buf = Vec::new();
        for r in remaining {
            buf.extend_from_slice(&Self::encode(r));
        }
        std::fs::write(&self.path, buf).ok();
    }

    /// WAL 자동 체크포인트가 필요한지 확인 (임계값 초과 여부)
    pub fn needs_auto_checkpoint(&self) -> bool {
        self.file_size() >= AUTO_CHECKPOINT_BYTES
    }
}

File: code/backend/core/src/transaction/wal.cpp
#include "engine/transaction/wal.hpp"

#include <cstdio>
#include <filesystem>
#include <fstream>
#include <stdexcept>

#include "engine/storage/atomic_write.hpp"

namespace fs = std::filesystem;

namespace engine {

namespace {
void push_u64_le(std::vector<std::uint8_t>& buf, std::uint64_t v) {
    for (int i = 0; i < 8; i++) buf.push_back(static_cast<std::uint8_t>((v >> (8 * i)) & 0xFF));
}
void push_u32_le(std::vector<std::uint8_t>& buf, std::uint32_t v) {
    for (int i = 0; i < 4; i++) buf.push_back(static_cast<std::uint8_t>((v >> (8 * i)) & 0xFF));
}
std::optional<std::uint64_t> read_u64_le(const std::vector<std::uint8_t>& buf, std::size_t pos) {
    if (pos + 8 > buf.size()) return std::nullopt;
    std::uint64_t v = 0;
    for (int i = 0; i < 8; i++) v |= static_cast<std::uint64_t>(buf[pos + static_cast<std::size_t>(i)]) << (8 * i);
    return v;
}
std::optional<std::uint32_t> read_u32_le(const std::vector<std::uint8_t>& buf, std::size_t pos) {
    if (pos + 4 > buf.size()) return std::nullopt;
    std::uint32_t v = 0;
    for (int i = 0; i < 4; i++) v |= static_cast<std::uint32_t>(buf[pos + static_cast<std::size_t>(i)]) << (8 * i);
    return v;
}
std::optional<std::string> read_string(const std::vector<std::uint8_t>& buf, std::size_t& pos) {
    auto len_opt = read_u32_le(buf, pos);
    if (!len_opt) return std::nullopt;
    std::size_t len = *len_opt;
    pos += 4;
    if (pos + len > buf.size()) return std::nullopt;
    std::string s(reinterpret_cast<const char*>(buf.data() + pos), len);
    pos += len;
    return s;
}
bool read_all_bytes(const std::string& path, std::vector<std::uint8_t>& out) {
    std::ifstream f(path, std::ios::binary);
    if (!f) return false;
    f.seekg(0, std::ios::end);
    auto size = f.tellg();
    if (size < 0) return false;
    out.resize(static_cast<std::size_t>(size));
    f.seekg(0);
    f.read(reinterpret_cast<char*>(out.data()), size);
    return true;
}
} // namespace

std::optional<WalOp> wal_op_from_u8(std::uint8_t v) {
    switch (v) {
        case 0x01: return WalOp::Insert;
        case 0x02: return WalOp::Update;
        case 0x03: return WalOp::Delete;
        case 0xFF: return WalOp::Commit;
        case 0xFE: return WalOp::Rollback;
        case 0xFD: return WalOp::Checkpoint;
        default: return std::nullopt;
    }
}

std::uint64_t TxnIoShared::next_id() {
    std::lock_guard<std::mutex> g(counter_mutex_);
    return next_txn_id_++;
}

WalManager::WalManager(const std::string& dir, std::shared_ptr<TxnIoShared> io)
    : path_(dir + "/rusql.wal"), io_(std::move(io)) {}

std::vector<std::uint8_t> WalManager::encode(const WalRecord& record) {
    std::vector<std::uint8_t> buf;
    buf.push_back(static_cast<std::uint8_t>(record.op));
    push_u64_le(buf, record.txn_id);
    push_u32_le(buf, static_cast<std::uint32_t>(record.table_name.size()));
    buf.insert(buf.end(), record.table_name.begin(), record.table_name.end());
    push_u32_le(buf, static_cast<std::uint32_t>(record.key.size()));
    buf.insert(buf.end(), record.key.begin(), record.key.end());
    push_u32_le(buf, static_cast<std::uint32_t>(record.data.size()));
    buf.insert(buf.end(), record.data.begin(), record.data.end());
    return buf;
}

std::optional<WalRecord> WalManager::decode(const std::vector<std::uint8_t>& buf, std::size_t& pos) {
    if (pos >= buf.size()) return std::nullopt;
    auto op = wal_op_from_u8(buf[pos]);
    if (!op) return std::nullopt;
    pos += 1;

    auto txn_id = read_u64_le(buf, pos);
    if (!txn_id) return std::nullopt;
    pos += 8;

    auto table_name = read_string(buf, pos);
    if (!table_name) return std::nullopt;
    auto key = read_string(buf, pos);
    if (!key) return std::nullopt;
    auto data = read_string(buf, pos);
    if (!data) return std::nullopt;

    return WalRecord{*op, *txn_id, *table_name, *key, *data};
}

// std::fstream has no portable fsync, but its underlying FILE*/fd does -- write via C
// stdio instead so a real disk sync (matching Rust's file.sync_all()) is possible. Every
// step's return value is checked and throws on failure, matching Rust's
// `.expect("WAL 기록 실패")`/`.expect("WAL fsync 실패")` panics -- a silently-swallowed
// write/flush/sync failure here would let a COMMIT report success without the record
// actually being durable, defeating the entire point of this function.
void WalManager::write_encoded_locked(const std::vector<std::uint8_t>& encoded, bool sync) const {
    std::FILE* fp = std::fopen(path_.c_str(), "ab");
    if (!fp) throw std::runtime_error("WAL 파일 열기 실패");
    std::size_t written = std::fwrite(encoded.data(), 1, encoded.size(), fp);
    if (written != encoded.size()) {
        std::fclose(fp);
        throw std::runtime_error("WAL 기록 실패");
    }
    if (std::fflush(fp) != 0) {
        std::fclose(fp);
        throw std::runtime_error("WAL 기록 실패");
    }
    if (sync) {
#ifdef _WIN32
        bool ok = _commit(_fileno(fp)) == 0;
#else
        bool ok = fsync(fileno(fp)) == 0;
#endif
        if (!ok) {
            std::fclose(fp);
            throw std::runtime_error("WAL fsync 실패");
        }
    }
    std::fclose(fp);
}

void WalManager::append(const WalRecord& record) {
    std::lock_guard<std::mutex> g(io_->wal_lock);
    write_encoded_locked(encode(record), false);
}

void WalManager::log_insert(std::uint64_t txn_id, const std::string& table, const std::string& key, const std::string& data) {
    append(WalRecord{WalOp::Insert, txn_id, table, key, data});
}

void WalManager::log_update(std::uint64_t txn_id, const std::string& table, const std::string& key, const std::string& data) {
    append(WalRecord{WalOp::Update, txn_id, table, key, data});
}

void WalManager::log_delete(std::uint64_t txn_id, const std::string& table, const std::string& key) {
    append(WalRecord{WalOp::Delete, txn_id, table, key, ""});
}

void WalManager::log_commit(std::uint64_t txn_id) {
    WalRecord record{WalOp::Commit, txn_id, "", "", ""};
    std::lock_guard<std::mutex> g(io_->wal_lock);
    write_encoded_locked(encode(record), true);
}

void WalManager::log_commit_no_sync(std::uint64_t txn_id) {
    WalRecord record{WalOp::Commit, txn_id, "", "", ""};
    std::lock_guard<std::mutex> g(io_->wal_lock);
    write_encoded_locked(encode(record), false);
}

void WalManager::log_rollback(std::uint64_t txn_id) {
    append(WalRecord{WalOp::Rollback, txn_id, "", "", ""});
}

void WalManager::log_checkpoint() {
    WalRecord record{WalOp::Checkpoint, 0, "", "", ""};
    std::lock_guard<std::mutex> g(io_->wal_lock);
    write_encoded_locked(encode(record), true);
}

std::vector<WalRecord> WalManager::read_all_locked() const {
    if (!fs::exists(path_)) return {};
    std::vector<std::uint8_t> buf;
    if (!read_all_bytes(path_, buf)) return {};
    std::vector<WalRecord> records;
    std::size_t pos = 0;
    while (auto record = decode(buf, pos)) records.push_back(*record);
    return records;
}

std::vector<WalRecord> WalManager::read_all() const {
    std::lock_guard<std::mutex> g(io_->wal_lock);
    return read_all_locked();
}

void WalManager::clear_locked() const {
    std::error_code ec;
    fs::remove(path_, ec);
}

void WalManager::clear() {
    std::lock_guard<std::mutex> g(io_->wal_lock);
    clear_locked();
}

void WalManager::remove_txn(std::uint64_t txn_id) {
    std::lock_guard<std::mutex> g(io_->wal_lock);
    auto all = read_all_locked();
    std::vector<WalRecord> remaining;
    for (auto& r : all) {
        if (r.txn_id != txn_id) remaining.push_back(r);
    }
    if (remaining.empty()) {
        clear_locked();
        return;
    }
    std::vector<std::uint8_t> buf;
    for (auto& r : remaining) {
        auto enc = encode(r);
        buf.insert(buf.end(), enc.begin(), enc.end());
    }
    write_bytes_atomic(path_, buf.data(), buf.size());
}

std::uint64_t WalManager::file_size() const {
    std::error_code ec;
    auto sz = fs::file_size(path_, ec);
    return ec ? 0 : static_cast<std::uint64_t>(sz);
}

void WalManager::truncate_to_last_checkpoint() {
    std::lock_guard<std::mutex> g(io_->wal_lock);
    auto records = read_all_locked();
    if (records.empty()) return;

    std::optional<std::size_t> last_cp;
    for (std::size_t i = records.size(); i-- > 0;) {
        if (records[i].op == WalOp::Checkpoint) { last_cp = i; break; }
    }
    if (!last_cp) return;

    std::vector<WalRecord> remaining(records.begin() + static_cast<std::ptrdiff_t>(*last_cp), records.end());
    if (remaining.size() <= 1) {
        clear_locked();
        return;
    }
    std::vector<std::uint8_t> buf;
    for (auto& r : remaining) {
        auto enc = encode(r);
        buf.insert(buf.end(), enc.begin(), enc.end());
    }
    write_bytes_atomic(path_, buf.data(), buf.size());
}

bool WalManager::needs_auto_checkpoint() const {
    return file_size() >= AUTO_CHECKPOINT_BYTES;
}

} // namespace engine
